File: 0063.unqiue_paths_II/solution.go
package leetcode0063

func uniquePathsWithObstacles(obstacleGrid [][]int) int {
	if obstacleGrid[0][0] == 1 {
		return 0
	}
	down := len(obstacleGrid)
	right := len(obstacleGrid[0])

	obstacleGrid[0][0] = 1

	for i := 1; i < right; i++ {
		if obstacleGrid[0][i] == 1 {
			obstacleGrid[0][i] = 0
		} else {
			obstacleGrid[0][i] = obstacleGrid[0][i-1]
		}
	}

	for i := 1; i < down; i++ {
		if obstacleGrid[i][0] == 1 {
			obstacleGrid[i][0] = 0
		} else {
			obstacleGrid[i][0] = obstacleGrid[i-1][0]
		}
	}
	for i := 1; i < down; i++ {
		for j := 1; j < right; j++ {
			if obstacleGrid[i][j] == 1 {
				obstacleGrid[i][j] = 0
			} else {
				obstacleGrid[i][j] = obstacleGrid[i-1][j] + obstacleGrid[i][j-1]
			}
		}
	}

	return obstacleGrid[down-1][right-1]
}

func main() {
	first := []int{0, 0, 1, 0}
	second := []int{0, 0, 0, 0}
	third := []int{0, 1, 0, 0}
	p := [][]int{first, second, third}
	print(uniquePathsWithObstacles(p) == 4)
}

File: 0063.unqiue_paths_II/solution.py
from typing import List


class Solution:
    def unique_paths_with_obstacle(self, obstacle_grid: List[List[int]]) -> int:
        """
        二维数组 关键点在于 obstacle_grid[i][j] = obstacle_grid[i-1][j] + obstacle_grid[i][j-1]
        如果遇到障碍1，后续都为0
        """
        if obstacle_grid[0][0]:  # 出发点是障碍
            return 0

        down = len(obstacle_grid)
        right = len(obstacle_grid[0])

        obstacle_grid[0][0] = 1

        # 判断第一行的情况
        for i in range(1, right):
            obstacle_grid[0][i] = obstacle_grid[0][i - 1] if obstacle_grid[0][i] == 0 else 0

        # 判断第一列的情况
        for i in range(1, down):
            obstacle_grid[i][0] = obstacle_grid[i - 1][0] if obstacle_grid[i][0] == 0 else 0

        # 从(1,1)开始将上和左的次数相加
        for i in range(1, down):
            for j in range(1, right):
                if obstacle_grid[i][j] == 1:
                    obstacle_grid[i][j] = 0
                else:
                    obstacle_grid[i][j] = obstacle_grid[i][j - 1] + obstacle_grid[i - 1][j]

        return obstacle_grid[down - 1][right - 1]


if __name__ == "__main__":
    grid = [[0, 0, 0], [0, 1, 0], [0, 0, 0]]
    assert Solution().unique_paths_with_obstacle(grid) == 2

    grid = [[0, 0, 0, 1], [0, 1, 0, 0], [0, 0, 0, 0]]
    assert Solution().unique_paths_with_obstacle(grid) == 3
